### DISCUSSION TITLE
What does strategy and size mean inside options object?

### DISCUSSION THREAD

**Discussion by marcelo-earth** [nasa/openmct#5746] [Community Q&A] [answered]
Hi OpenMCT community!

I'm working on a simple historical telemetry provider. And I see that when retrieving historical data, I have access to an `options` object from the parameters:

```js
openmct.telemetry.addProvider({
    ...
    request: function (domainObject, options) {
        console.log({ options });
    },
    ...
});
```

My question is, inside the `options` object I have access to a `strategy` and `size` properties, what do those properties mean?

```json
{
    "options": {
        "size": 222,
        "strategy": "minmax",
        "start": 1662821170502,
        "end": 1662822070502,
        "domain": "utc",
        "signal": {}
    }
}
```

**Comment by akhenry** [comment, marked as answer]
Hi there!

Good question. These are both optional arguments that some of our views pass through to allow telemetry providers to optimize certain requests. 

The `size` argument tells the telemetry provider the maximum number of results being requested. It is used most commonly in two cases:
1. "Latest" type views (such as the Latest Available Data or LAD Table) can only show one telemetry value at a time, so fetching more than that is a waste of time. So, it instructs the telemetry provider to only fetch one value if the server supports that sort of request.
2. Plots will always provide the "size" option as well, which will be equal to the number of horizontal pixels available to the plot. Returning any more values than this from the server is a waste of time because the plot doesn't have the pixels available to display them. Again, it's optional, because the server may not necessarily support this.

The `strategy` argument is somewhat related to `size`. There are two commonly used strategies in our codebase:
1. "Latest" type views will make requests using a strategy of `latest`. This tells the telemetry provider that the view is only interested in the very latest telemetry value for the given point. This is always paired with a `size` of 1. Some servers provide different API endpoints for historical and latest requests because historical values may be fetched from a database, while latest values are just kept in memory, so we separate `size` and `strategy` to draw that distinction.
2. Plots will mostly make historical requests with a `strategy` of `minmax`. The `minmax` strategy is used by servers as a way of summarizing the telemetry values for a given time period so that the server does not have to send every single value within the request time period. In this strategy the time period being requested is divided up into equally sized buckets, and then the minimum and maximum value within each of those buckets is returned, instead of returning all of the telemetry values. These are sometimes referred to as "samples" (as in the Yamcs telemetry server), or "rollups". Most telemetry servers we work with support some version of this. This is paired with the `size` parameter, which determines the maximum number of buckets to return to Open MCT.

These are also covered in [our API documentation](https://github.com/nasa/openmct/blob/master/API.md#request-strategies-draft).

Hope this helps to clarify things!

**Reply by marcelo-earth** [reply]
Hi @akhenry!
Thank you very much for your response. One last question about that! 🙏

> These are both optional arguments that **some of our views pass through** to allow telemetry providers to optimize certain requests.

I would like to test `latest` strategy. What views I could use to test this? I'm using the spacecraft example from [openmct-tutorial](https://github.com/nasa/openmct-tutorial) as a guide. I couldn't find how to test `latest` strategy here.

<img width="1233" alt="image" src="https://user-images.githubusercontent.com/47821093/189795710-ce5858d3-dc63-4ea4-a03b-35c26e733c05.png">

**Reply by jvigliotta** [reply]
@360macky You can test this by using our LAD Tables (Latest Available Data). These will pass in the 'latest' for the strategy option.

**Reply by marcelo-earth** [reply]
Thanks @jvigliotta! I'm checking that option.
